algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py

```diff
@@ -0,0 +1,253 @@
+                        split: str,
+                        data_dir: str,
+                        global_batch_size: int,
+                        num_batches: Optional[int] = None,
+                        repeat_final_dataset: bool = False):
+    per_device_batch_size = int(global_batch_size / N_GPUS)
+    n_inputs = 4
+    keys = ['inputs', 'input_paddings', 'targets', 'target_paddings']
+    np_iter = super().build_input_queue(data_rng,
+                                        split,
+                                        data_dir,
+                                        global_batch_size,
+                                        num_batches,
+                                        repeat_final_dataset)
+    while True:
+      # Only iterate over tf input pipeline in one Python process to
+      # avoid creating too many threads.
+      if RANK == 0:
+        batch = next(np_iter)  # pylint: disable=stop-iteration-return
+        tensor_list = []
+        for key in keys:
+          value = batch[key]
+          tensor = torch.as_tensor(value, dtype=torch.float32, device=DEVICE)
+          tensor_list.append(tensor)
+          batch[key] = (
+              tensor[0] if USE_PYTORCH_DDP else tensor.view(
+                  -1, *value.shape[2:]))
+        # Send batch to other devices when using DDP.
+        if USE_PYTORCH_DDP:
+          # During eval, the batch size of the remainder might be different.
+          if split != 'train':
+            per_device_batch_size = torch.tensor(
+                len(batch['inputs']), dtype=torch.int32, device=DEVICE)
+            dist.broadcast(per_device_batch_size, src=0)
+          dist.broadcast(torch.cat(tensor_list, dim=-1), src=0)
+      else:
+        # During eval, the batch size of the remainder might be different.
+        if split != 'train':
+          per_device_batch_size = torch.empty((1,),
+                                              dtype=torch.int32,
+                                              device=DEVICE)
+          dist.broadcast(per_device_batch_size, src=0)
+        tensor = torch.empty(
+            (N_GPUS, per_device_batch_size, 320000 * 2 + 256 * 2),
+            dtype=torch.float32,
+            device=DEVICE)
+        dist.broadcast(tensor, src=0)
+        # Note that the order of the keys is important.
+        tensors = tensor.split([320000, 320000, 256, 256], dim=-1)
+        batch = {}
+        for key, n in zip(keys, range(n_inputs)):
+          batch[key] = tensors[n][RANK]
+      yield batch
+
+  def compute_loss(self, logits, logit_paddings, targets, target_paddings):
+    logprobs = torch.log_softmax(logits, dim=-1)
+    input_lengths = torch.einsum('bh->b', 1 - logit_paddings).long()
+    target_lengths = torch.einsum('bh->b', 1 - target_paddings).long()
+    per_seq_loss = self.ctc_loss(
+        logprobs.permute(1, 0, 2),
+        targets.long(),
+        input_lengths,
+        target_lengths)
+    average_loss = per_seq_loss / target_lengths.sum()
```

Comment:
(we need to do a pass over the codebase to standardize this sort of logic, I know we do this in the jax workload, so this is an issue with both, but I'll talk to Sourabh about if I'm mistaken)

instead of returning the average here, do we instead need to make a helper function that returns the per_seq_loss and summed target len, and then inside `_eval_model_on_split` we would call the new helper function and sum up the numerator and denominator before dividing? (then `loss_fn` would just call the helper and immediately return the division)

Reply:
(I also see that I have set `reduction='none'` for the ctc-loss and haven't summed the `per_seq_loss` before computing the average_loss -- will fix this!)